src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -717,6 +717,9 @@ private Task ProcessSettingsFrameAsync(in ReadOnlySequence<byte> payload)
 
                 _clientSettings.Update(Http2FrameReader.ReadSettings(payload));
 
+                // Maximum encoder size is limited by the configured max on the server
+                _frameWriter.UpdateMaxHeaderTableSize(Math.Min(_clientSettings.HeaderTableSize, (uint)Limits.Http2.HeaderTableSize));
```

Comment:
I think this should be done after we write the SETTINGS ACK.

Reply:
To that point, I think updating the size and sending the ack need to be atomic/locked. We don't want responses sent in between the two operations regardless of the ordering since the settings would be temporarily out of sync.

Reply:
_frameWriter already takes the connection-wide _writeLock in UpdateMaxHeaderTableSize and WriteSettingsAckAsync.

Reply:
Right, but it releases the lock between the two calls. I'm saying that both need to happen in the same lock.

Reply:
> We don't want responses sent in between the two operations regardless of the ordering since the settings would be temporarily out of sync.

I think happening afterwards is fine and a lock isn't required. Decoders don't care about the ACK, they use the dynamic table size update written in headers to control the size of their dynamic table.

For example, today Chrome sends 64 * 1024 as SETTINGS_HEADER_TABLE_SIZE. Because we never respond with a new size in the dynamic table header, Chrome doesn't change its table size, even though it got an ACK.

Reply:
Good point, I'd forgotten about the resize built into the hpack commands.